### PR TITLE
#8: Paper tasks using puppeteer (closes #8)

### LINKED ISSUE #8: Paper tasks using puppeteer

### PR DESCRIPTION
**Pull request by ecamellini**
Closes #8

### tests performed

image

### CODE REVIEW

File: paper-tasks-puppeteer/paper-tasks-puppeteer.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  // count tasks
+  const tasks = await page.$$('.hp-task')
+  return tasks.length
+}
+
+function onSuccess(tasksCount) {
+  console.log(tasksCount)
+  console.log('---')
+  console.log('Open in Chrome | href=https://paper.dropbox.com/tasks')
```
Comment (gabro) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.js]: Open in browser

File: paper-tasks-puppeteer/paper-tasks-puppeteer.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  browser.close()
+}
+
+// close browser on CTRL+C
+process.on('SIGINT', () => {
+  browser.close()
+})
+
+// run script
+getPaperTasksCount()
```
Comment (gabro) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.js]: I would be great to also retrieve the list of tasks and show it directly in bitbar. How hard is it?

**Comment by FrancescoCioria** [comment]
@ecamellini I fixed the login (you weren't using the input's selector), now it should work
NOTE: I refactored other parts of the code while looking for the problem

File: paper-tasks-puppeteer/paper-tasks-puppeteer.js (hunk 2 above)
Comment (FrancescoCioria) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.js]: not much, I already implemented the logic in a previous commit

Comment (FrancescoCioria) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.js]: https://github.com/buildo/bitbar-plugins/pull/9/commits/3cb27131d37aa1b36fdfe35778ee8313f8a256a7#diff-01b0f3911ea89d6090ff9b02cdd694aaR40

I haven't tested it though

**Comment by ecamellini** [comment]
@FrancescoCioria @gabro fixed everything needed to make it work in the BitBar env. Tested!

File: paper-tasks-puppeteer/paper-tasks-puppeteer.2m.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env /usr/local/bin/node
```
Comment (FrancescoCioria) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.2m.js]: NOTE: I had to change this too to:
```
/Users/francesco/.nvm/versions/v8.9.4/bin/node
```

Comment (ecamellini) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.2m.js]: Yep I put it in the README

**Comment by FrancescoCioria** [comment]
@ecamellini @gabro with the last commit we're showing each task in the dropdown, do you like it?

image

NOTE: when you click on the you always finish in the general page `https://paper.dropbox.com/tasks`

**Comment by ecamellini** [comment]
👏

**Comment by gabro** [comment]
Pretty cool! Can we pick up the due date too?

**Comment by gabro** [comment]
Also, how hard is it to get the exact document URL? Finally I would add a separator between the last task and the "open in browser" line

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env /usr/local/bin/node
+
+
+/******** CONFIGURATION ********/
+const email = '...' // e.g., francesco@buildo.io
+const password = '...' // your dropbox password
+const puppeteerPath = '...' // e.g., /usr/local/lib/node_modules/puppeteer
```
Comment (gabro) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: Why?

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  browser.close()
+}
+
+function onError(e) {
+  console.error(e)
+  browser.close()
+}
+
+// close browser on CTRL+C
+process.on('SIGINT', () => {
```
Comment (gabro) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: What is this for?

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+const puppeteerPath = '...' // e.g., /usr/local/lib/node_modules/puppeteer
+/*************** ***************/
+
+
+const puppeteer = require(puppeteerPath)
+
+// global variable is needed to call "browser.close()" when script is over
+let browser
+
+async function getPaperTasksCount() {
```
Comment (gabro) [inline paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: getPaperTasks

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js (hunk 5 above)
Comment (FrancescoCioria) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: come on, there's even a comment! 😝

**Comment by FrancescoCioria** [comment]
> Pretty cool! Can we pick up the due date too?

we could (but I won't :P)

> Also, how hard is it to get the exact document URL?

there's no link, so very hard

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js (hunk 4 above)
Comment (FrancescoCioria) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: bitbar limitation, it needs the exact path to `node` executable and to every global library

File: paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js (hunk 5 above)
Comment (gabro) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: Why do you ever need to press CTRL-C?

Comment (FrancescoCioria) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: when testing I found myself pressing CTRL+C often, in real life you probably don't need it, but still it shouldn't hurt

Comment (gabro) [reply paper-tasks-puppeteer/paper-tasks-puppeteer.30m.js]: 👍 I was just wondering

**Comment by gabro** [comment]
I've improved this a bit:

- group tasks by paper
- link to the specific task in the paper
- due date
- overdue dates are displayed in red

image

**Pull request merged**